streaming/stream_transfer_task.cc

```diff
@@ -237,21 +237,28 @@ future<> stream_transfer_task::execute() {
     }).then([this, id, plan_id] {
         _mutation_done_sent = true;
         sslog.debug("[Stream #{}] GOT STREAM_MUTATION_DONE Reply from {}", plan_id, id.addr);
-    }).handle_exception([this, plan_id, cf_id, id] (std::exception_ptr ep) {
+    }).handle_exception(coroutine::lambda([this, plan_id, cf_id, id] (std::exception_ptr ep) -> future<> {
         // If the table is dropped during streaming, we can ignore the
         // errors and make the stream successful. This allows user to
         // drop tables during node operations like decommission or
         // bootstrap.
-        if (!session->manager().db().column_family_exists(cf_id)) {
-            sslog.warn("[Stream #{}] Ignore the table with table_id {} which is dropped during streaming: {}", plan_id, cf_id, ep);
-            if (!_mutation_done_sent) {
-                return session->manager().ms().send_stream_mutation_done(id, plan_id, _ranges, cf_id, session->dst_cpu_id);
-            }
-            return make_ready_future<>();
+        //
+        // the db table metadata on different shards are not necessarily in
+        // sync, but if the table is dropped on any one of them, the exception
+        // is thrown. so we need to check on all shards.
+        bool cf_exists = co_await session->manager().db().container().map_reduce0(
+            [cf_id](const replica::database& db) { return db.column_family_exists(cf_id); },
+            true, std::logical_and<bool>());
+        if (cf_exists) {
+            sslog.warn("[Stream #{}] stream_transfer_task: Fail to send to {}: {}", plan_id, id, ep);
+            co_await coroutine::return_exception_ptr(ep);
```

Comment:
Why not just call:

std::rethrow_exception(ep);

Reply:
thank you for pointing this out. i always have a hard time to tell when to use `coroutine::return_exception_ptr()` and when to use `std::rethrow_exception()`.

Reply:
`coroutine::return_exception_ptr()` only works in coroutines. It is more optimal version of `std::rethrow_exception()`, because unlike the latter, the former does not involve actually throwing and re-catching the exception.

Reply:
as always, thank you Botond! i will try to improve Seastar's document on this. i should not be the only one who is confused.

Reply:
https://github.com/scylladb/seastar/pull/2086